src/flow/max-flow-min-cut.jl

```diff
@@ -0,0 +1,135 @@
+            sizehint!(path, n)
+
+            u = v
+            while u!=source                # trace path from v to source
+                u = P[u]
+                push!(path, u)
+            end
+            reverse!(path)
+
+            u = v                          # trace path from v to target
+            while u!=target
+                u = S[u]
+                push!(path, u)
+            end
+                                           # augment flow along path
+            flow += augment_path!(path, flow_matrix, capacity_matrix)
+        end
+    end
+
+    return flow, flow_matrix
+end
+
+"""Calculates the amount by which flow can be augmented in the given path.
+Augments the flow and returns the augment value."""
+
+function augment_path!{T<:Number}(
+    path::Vector{Int},                     # input path
+    flow_matrix::AbstractArray{T,2},       # the current flow matrix
+    capacity_matrix::AbstractArray{T,2}    # edge flow capacities
+    )
+    augment = typemax(T)                   # initialize augment
+    for i in 1:length(path)-1              # calculate min capacity along path
+        u = path[i]
+        v = path[i+1]
+        augment = min(augment,capacity_matrix[u,v] - flow_matrix[u,v])
+    end
+
+    for i in 1:length(path)-1              # augment flow along path
+        u = path[i]
+        v = path[i+1]
+        flow_matrix[u,v] += augment
+        flow_matrix[v,u] -= augment
+    end
+
+    return augment
+end
+
+"""Uses Bidirectional BFS to look for augmentable-paths. Returns the vertex where
+the two BFS searches intersect, the Parent table of the path as well as the
+Successor table of the path found."""
+
+function fetch_path{T<:Number}(
+    flow_graph::LightGraphs.DiGraph,       # the input graph
+    source::Int,                           # the source vertex
+    target::Int,                           # the target vertex
+    flow_matrix::AbstractArray{T,2},       # the current flow matrix
+    capacity_matrix::AbstractArray{T,2}    # edge flow capacities
+    )
+    n = nv(flow_graph)
+
+    P = [-1 for i in 1:n]                  # parent table of path
+    P[source] = -2
+
+    S = [-1 for i in 1:n]                  # successor table of path
+    S[target] = -2
+
+    Q_f = [source]                         # forward queue
+    sizehint!(Q_f, n)
+
+    Q_r = [target]                         # reverse queue
+    sizehint!(Q_r, n)
+
+    while true
+
+        if length(Q_f) <= length(Q_r)
+            u = pop!(Q_f)
+            for v in fadj(flow_graph, u)
+                if capacity_matrix[u,v] - flow_matrix[u,v] > 0 && P[v] == -1
+                    P[v] = u
+                    if S[v] == -1
+                        unshift!(Q_f, v)
+                    else
+                        return v, P, S
+                    end
+                end
+            end
+
+            if length(Q_f) == 0
+                return 0, nothing, nothing # No paths to target
```

Comment:
Per previous comment by @jpfairbanks, this is type unstable. It's much better to ensure you return the same types of values from a function regardless of outcome.